chacha20_poly1305/src/poly1305.rs

```diff
@@ -0,0 +1,274 @@
+// SPDX-License-Identifier: CC0-1.0
+
+//! Poly1305 one-time message authenticator from RFC8439.
+//!
+//! Heavily inspired by the ["Donna"](https://github.com/floodyberry/poly1305-donna/blob/master/poly1305-donna-32.h) implementation in C
+//! and Loup Vaillant's [Poly1305 design article](https://loup-vaillant.fr/tutorials/poly1305-design).
+
+/// 2^26 for the 26-bit limbs.
+const BITMASK: u32 = 0x03ffffff;
+/// Number is encoded in five 26-bit limbs.
+const CARRY: u32 = 26;
+
+/// Poly1305 authenticator takes a 32-byte one-time key and a message and produces a 16-byte tag.
+///
+/// 64-bit constant time multiplication and addition implementation.
+#[derive(Debug)]
+pub struct Poly1305 {
+    /// r part of the secret key.
+    r: [u32; 5],
+    /// s part of the secret key.
+    s: [u32; 4],
+    /// State used to create tag.
+    acc: [u32; 5],
+    /// Leftovers between adds.
+    leftovers: [u8; 16],
+    /// Track relevant leftover bytes.
+    leftovers_len: usize,
+}
+
+impl Poly1305 {
+    /// Initialize authenticator with a 32-byte one-time secret key.
+    pub const fn new(key: [u8; 32]) -> Self {
+        // Taken from Donna. Assigns r to a 26-bit 5-limb number while simultaneously 'clamping' r.
+        // Casts are used to keep this constructor constant.
+        let r0 =
+            (key[0] as u32 | (key[1] as u32) << 8 | (key[2] as u32) << 16 | (key[3] as u32) << 24)
+                & 0x3ffffff;
+        let r1 = ((key[3] as u32) >> 2
+            | (key[4] as u32) << 6
+            | (key[5] as u32) << 14
+            | (key[6] as u32) << 22)
+            & 0x03ffff03;
+        let r2 = ((key[6] as u32) >> 4
+            | (key[7] as u32) << 4
+            | (key[8] as u32) << 12
+            | (key[9] as u32) << 20)
+            & 0x03ffc0ff;
+        let r3 = ((key[9] as u32) >> 6
+            | (key[10] as u32) << 2
+            | (key[11] as u32) << 10
+            | (key[12] as u32) << 18)
+            & 0x03f03fff;
+        let r4 = ((key[12] as u32) >> 8
+            | (key[13] as u32)
+            | (key[14] as u32) << 8
+            | (key[15] as u32) << 16)
+            & 0x000fffff;
+
+        let s0 = key[16] as u32
+            | (key[17] as u32) << 8
+            | (key[18] as u32) << 16
+            | (key[19] as u32) << 24;
+        let s1 = key[20] as u32
+            | (key[21] as u32) << 8
+            | (key[22] as u32) << 16
+            | (key[23] as u32) << 24;
+        let s2 = key[24] as u32
+            | (key[25] as u32) << 8
+            | (key[26] as u32) << 16
+            | (key[27] as u32) << 24;
+        let s3 = key[28] as u32
+            | (key[29] as u32) << 8
+            | (key[30] as u32) << 16
+            | (key[31] as u32) << 24;
```

Comment:
This is pretty hard to read `u32::from_le_bytes` would be definitely better.

Reply:
For sure, it would be an improvement, but I don't think this is hard to read. It looks like very standard byte-marshalling code to me.

Reply:
I mean it's hard to review. When I read `from_le_bytes` I know exactly what it does, when I read this then I have to look at al the numbers and remember which position needs to be which number etc.

Reply:
I switched over to a `from_le_bytes` version. I had that at some point, but dropped it when I thought the MSRV was still 1.56.